enclave-tls/src/crypto_wrappers/openssl/gen_cert.c

```diff
@@ -0,0 +1,167 @@
+{
+	int nid;
+	ASN1_OCTET_STRING *octet = NULL;
+	X509_EXTENSION *ext = NULL;
+	int ret = -1;
+
+	nid = OBJ_create(oid, NULL, NULL);
+	if (nid == NID_undef) {
+		ETLS_DEBUG("obj create failed, %s\n", oid);
+		return ret;
+	}
+
+	octet = ASN1_OCTET_STRING_new();
+	if (!octet)
+		goto err;
+
+	ASN1_OCTET_STRING_set(octet, data, data_len);
+
+	ext = X509_EXTENSION_create_by_NID(NULL, nid, 0, octet);
+	if (!ext) {
+		ETLS_DEBUG("extension create failed, %s\n", oid);
+		goto err;
+	}
+
+	if (!X509_add_ext(cert, ext, -1)) {
+		ETLS_DEBUG("extension add failed, %s\n", oid);
+		goto err;
+	}
+
+	ret = 0;
+
+err:
+	if (ret == -1)
+		ETLS_DEBUG("X509 extension add failed, %s, nid = %d\n", oid, nid);
+
+	if (ext)
+		X509_EXTENSION_free(ext);
+
+	if (octet)
+		ASN1_OCTET_STRING_free(octet);
+
+	return ret;
+}
+
+crypto_wrapper_err_t openssl_gen_cert(crypto_wrapper_ctx_t *ctx,
+ 			       enclave_tls_cert_info_t *cert_info)
+{
+	struct openssl_ctx *octx;
+	cert_subject_t *subject;
+	X509 *cert = NULL;
+	X509_NAME *name;
+	EVP_PKEY *pkey = NULL;
+	unsigned char *der;
+	int len;
+	int ret;
+
+	ETLS_DEBUG("ctx %p, cert_info %p\n", ctx, cert_info);
+
+	if (!ctx || !cert_info)
+		return -CRYPTO_WRAPPER_ERR_INVALID;
+
+	octx = ctx->crypto_private;
+
+	pkey = EVP_PKEY_new();
+	if (!pkey)
+		return -CRYPTO_WRAPPER_ERR_NO_MEM;
+
+	ret = -CRYPTO_WRAPPER_ERR_PRIV_KEY_LEN;
+	if (!EVP_PKEY_assign_RSA(pkey, octx->key))
+		goto err;
+
+	ret = -CRYPTO_WRAPPER_ERR_NO_MEM;
+	cert = X509_new();
+	if (!cert)
+		goto err;
+
+	X509_set_version(cert, 2);
+	ASN1_INTEGER_set(X509_get_serialNumber(cert), 9527);
+	X509_gmtime_adj(X509_get_notBefore(cert), 0);
+	/* 10 years */
+	X509_gmtime_adj(X509_get_notAfter(cert), 3600 * 24 * 365 * 10);
+
+	ret = -CRYPTO_WRAPPER_ERR_PUB_KEY_LEN;
+	if (!X509_set_pubkey(cert, pkey))
+		goto err;
+
+	/* subject name */
+	name = X509_get_subject_name(cert);
+	if (!name)
+		goto err;
+
+	subject = &cert_info->subject;
+	X509_NAME_add_entry_by_txt(name, "O", MBSTRING_ASC,
+			subject->organization, -1, -1, 0);
+	X509_NAME_add_entry_by_txt(name, "OU", MBSTRING_ASC,
+			subject->organization_unit, -1, -1, 0);
+	X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC,
+			subject->common_name, -1, -1, 0);
+	if (!X509_set_issuer_name(cert, name))
+		goto err;
+
+	ret = -CRYPTO_WRAPPER_ERR_PUB_KEY_DECODE;
+
+	ETLS_DEBUG("evidence type '%s' requested\n", cert_info->evidence.type);
+
+	if (strcmp(cert_info->evidence.type, "sgx_epid") == 0) {
+		attestation_verification_report_t *epid = &cert_info->evidence.epid;
+
+		x509_extension_add(cert, "1.2.840.113741.1337.2",
+				epid->ias_report, epid->ias_report_len);
+		x509_extension_add(cert, "1.2.840.113741.1337.3",
+				epid->ias_sign_ca_cert, epid->ias_sign_ca_cert_len);
+		x509_extension_add(cert, "1.2.840.113741.1337.4",
+				epid->ias_sign_cert, epid->ias_sign_cert_len);
+		x509_extension_add(cert, "1.2.840.113741.1337.5",
+				epid->ias_report_signature, epid->ias_report_signature_len);
```

Comment:
add error handling for above functions.

Reply:
@YangLiang3 You need to discuss how to share your openssl.h to this instance for the common definitions, e.g, the OIDs.

Reply:
mark. global common header file will define OIDs.